### PR TITLE
:bug: stopped duplication of incidents while creating violations

### PR DESCRIPTION
**Pull request by Chanakya-TS**
fixes issue #144

### CODE REVIEW

File: engine/engine.go [hunk 1]
```diff
@@ -416,7 +418,17 @@ func (r *ruleEngine) createViolation(conditionResponse ConditionResponse, rule R
 			incident.Message = templateString
 		}
 
-		incidents = append(incidents, incident)
+		// Checking duplication of incident
+
+		jsonIncident, _ := json.Marshal(incident)  // Converting to JSON to be used as keys of a map
```
Comment (shawn-hurley) [inline engine/engine.go:423]: I am worried about the cost of this. Can we make a key function that only takes specific values to compare? 

Think like file URI, Message, and Linenumber if it exists in extras. I think that the cost of marshal could become too great, and the things that this is capturing as different issues, I think is misleading. 

For instance, if the type is a method or a local variable, as long as we are pointing to the same location, it is finding the same violation, just in a different search mechanism, and I think the user can be spared this detail IMO.

File: engine/engine.go [hunk 2]
```diff
@@ -416,7 +418,17 @@ func (r *ruleEngine) createViolation(conditionResponse ConditionResponse, rule R
...
+		stringJsonIncident := string(jsonIncident) // Converting JSON to string
+
+		_, isDuplicate := incidentsSet[stringJsonIncident] // Checking if incident already exists in set
```
Comment (shawn-hurley) [inline engine/engine.go:426]: I usually perfer to make this a one line, but this is just a nit.

```
if _, ok := ...... ; !ok {
  do x
  }
  ```

Comment (Chanakya-TS) [reply engine/engine.go:426]: > I usually perfer to make this a one line, but this is just a nit.
> 
> ```
> if _, ok := ...... ; !ok {
>   do x
>   }
> ```

ah okay, will do that from next time. thank you!

**Review by shawn-hurley — COMMENTED** [review]
I think this is a great PR, I do think that we may need to update the rule-example.

I believe that this is filtering out (now correctly 😄 ) the XML XPath on the pom file. 

One way to fix this is to update the `rule-example.yaml` to include a variable in the message.

**Review by pranavgaikwad — CHANGES_REQUESTED** [review]
Our CI expects content of the `container-output.yaml` to match the output after running analyzer with your changes. It looks like the change you made deduplicates a lot of the incidents our CI is expecting. Can you please  update the `container-output.yaml` to match the new expected output?

File: demo-output.yaml [hunk 3]
```diff
@@ -138,8 +138,76 @@
       - testing
       - test
       incidents:
+      - uri: file:///analyzer-lsp/cmd/analyzer/main.go
```
Comment (shawn-hurley) [inline demo-output.yaml:141]: These should not be added back, we got rid of these

Comment (Chanakya-TS) [reply demo-output.yaml:141]: thats weird, gimme a sec

Comment (Chanakya-TS) [reply demo-output.yaml:141]: i think something happened while rebasing

Comment (Chanakya-TS) [reply demo-output.yaml:141]: wait no, that is the output that i am getting when i run the build. im not sure what the problem is

**Review by shawn-hurley — APPROVED** [review]
I think this looks good I wonder if it will kill our integration numbers

**Review by pranavgaikwad — APPROVED** [review]

**Comment by pranavgaikwad** [comment]
I was just making sure we don't break integration test with this, and I found https://github.com/konveyor/analyzer-lsp/blob/2f7088e1a51b8ea40b7343fff679d43a972b9b44/provider/internal/builtin/provider.go#L187, I think we need safeguard for a non integer type line number

**Pull request merged**